apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,382 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc Reference implementation of the storage.
+%%
+%% Trivial, extremely slow and inefficient. It also doesn't handle
+%% restart of the Erlang node properly, so obviously it's only to be
+%% used for testing.
+-module(emqx_ds_storage_bitfield_lts).
+
+-behaviour(emqx_ds_storage_layer).
+
+%% API:
+-export([]).
+
+%% behavior callbacks:
+-export([create/4, open/5, store_batch/4, get_streams/4, make_iterator/5, next/4]).
+
+%% internal exports:
+-export([format_key/2, format_keyfilter/1]).
+
+-export_type([options/0]).
+
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+-type options() ::
+    #{
+        bits_per_wildcard_level => pos_integer(),
+        topic_index_bytes => pos_integer(),
+        epoch_bits => non_neg_integer()
+    }.
+
+%% Permanent state:
+-type schema() ::
+    #{
+        bits_per_wildcard_level := pos_integer(),
+        topic_index_bytes := pos_integer(),
+        ts_bits := non_neg_integer(),
```

Comment:
Nit. It seems that if we use `timestamp_bytes` here, the assertion in `make_keymapper` won't be needed.